src/Loan.ts

```diff
@@ -0,0 +1,143 @@
+
+export class Loan {
+    public readonly id: number;
+    public readonly collateralAmount: Wei;
+    public readonly borrower: string;
+    public readonly productId: number;
+    public readonly state: number;
+    public readonly maturity: number;
+    public readonly disbursementTime: number;
+    public readonly interestAmount: Tokens;
+    public readonly loanAmount: Tokens;
+    public readonly repaymentAmount: Tokens;
+    public readonly loanManagerAddress: string;
+    public readonly marginCallRate: Tokens;
+    public readonly tokenAddress: string;
+    private readonly _isCollectable: boolean;
+
+    constructor(loan: ILoanTuple, loanManagerAddress: string, tokenAddress: string) {
+        const [
+            configId,
+            configCollateralAmount,
+            configRepaymentAmount,
+            configBorrower,
+            configProductId,
+            configState,
+            configMaturity,
+            configDisbursementTime,
+            configLoanAmount,
+            configInterestAmount,
+            configMarginCallRate,
+            configIsCollectable
+        ]: ILoanTuple = loan;
+
+        this.loanManagerAddress = loanManagerAddress;
+        this.tokenAddress = tokenAddress;
+
+        const maturity:number = parseInt(configMaturity, 10);
+        if (maturity > 0) {
+            const disbursementTime:number = parseInt(configDisbursementTime, 10);
+
+            const state: number = parseInt(configState, 10);
+
+            this.id = parseInt(configId, 10);
+            this.collateralAmount = Wei.parse(configCollateralAmount);
+            this.borrower = configBorrower;
+            this.productId = parseInt(configProductId, 10);
+            this.state = state;
+            this.maturity = maturity;
+            this.disbursementTime = disbursementTime;
+            this.interestAmount = Tokens.parse(configInterestAmount);
+            this.loanAmount = Tokens.parse(configLoanAmount);
+            this.repaymentAmount = Tokens.parse(configRepaymentAmount);
+
+            if(configMarginCallRate !== undefined) {
+                this.marginCallRate = Tokens.parse(configMarginCallRate);
+                this._isCollectable = configIsCollectable === '1';
+            }
+        }
+    }
+
+
+    get isMarginLoan(): boolean {
+        return !!this.marginCallRate
+    }
+
+    get term(): number {
+        return this.maturity - this.disbursementTime;
+    }
+
+    get isRepayable(): boolean {
+        return this.state === LOAN_STATES.Open && currentTime() < this.maturity;
+    }
+
+    get isCollectable(): boolean {
+        return this.isMarginLoan ? this._isCollectable : this.state === LOAN_STATES.Defaulted;
+    }
+
+    get isDue(): boolean {
+        return this.state === LOAN_STATES.Open && (this.maturity - currentTime() < sevenDays);
+    }
+
+    // TODO: update for margin loans!!!
+    get loanStateText(): string {
```

Comment:
This belongs to the frontend display code.

Reply:
done. method was removed.